src/fluree/db/query/subject_crawl/subject.cljc

```diff
@@ -0,0 +1,111 @@
+                          :post [(flake/->Flake nil p o* nil nil -2147483647)
+                                 (flake/->Flake nil p o* nil nil 2147483647)]
+                          :psot [(flake/->Flake nil p nil nil nil -2147483647)
+                                 (flake/->Flake nil p nil nil nil 2147483647)])
+        idx-root    (get db idx)
+        cmp         (:comparator idx-root)
+        range-set   (flake/sorted-set-by cmp fflake lflake)
+        in-range?   (fn [node]
+                      (query-range/intersects-range? node range-set))
+        query-xf    (where-subj-xf {:start-test  >=
+                                    :start-flake fflake
+                                    :end-test    <=
+                                    :end-flake   lflake
+                                    ;; if looking for pred + obj, but pred is not indexed, then need to use :psot and filter for 'o' values
+                                    :xf          (when (and (= :psot idx) o*)
+                                                   (map (fn [flakes]
+                                                          (filter #(= o* (flake/o %)) flakes))))})
+        resolver    (index/->CachedTRangeResolver conn novelty t t (:async-cache conn))
+        tree-chan   (index/tree-chan resolver idx-root in-range? query-range/resolved-leaf? 1 query-xf error-ch)
+        return-chan (async/chan 10 (comp (map flake/s)
+                                         (dedupe)))]
+    (async/go-loop []
+      (let [next-chunk (<! tree-chan)]
+        (if (nil? next-chunk)
+          (async/close! return-chan)
+          (let [more? (loop [vs (seq next-chunk)
+                             i  0]
+                        (if vs
+                          (if (>! return-chan (first vs))
+                            (recur (next vs) (inc i))
+                            false)
+                          true))]
+            (if more?
+              (recur)
+              (async/close! return-chan))))))
+    return-chan))
+
+
+(defn flakes-xf
+  [{:keys [db fuel-vol max-fuel error-ch vars filter-map permissioned?] :as _opts}]
+  (let [permissions (when permissioned?
+                      (subj-perm-filter-fn db))]
+    (fn [sid port]
+      (async/go
+        (try*
+          ;; TODO: Right now we enforce permissions after the index-range call, but
+          ;; TODO: in some circumstances we can know user can see no subject flakes
+          ;; TODO: and if detected, could avoid index-range call entirely.
+          (let [flakes (cond->> (<? (query-range/index-range db :spot = [sid]))
+                                filter-map (filter-subject vars filter-map)
+                                permissioned? permissions
+                                permissioned? <?)]
+            (when (seq flakes)
+              (async/put! port flakes))
+
+            (async/close! port))
+          (catch* e (async/put! error-ch e) (async/close! port) nil))))))
```

Comment:
Is this the same as or very similar to `rdf-type/flakes-xf`? If so, should we move (parts) to a shared namespace and DRY up the code a little?

Reply:
Not without opening up new channels in the process which I want to avoid... one performs on individual subject-ids whereas the other performs on groups of flakes by subject.